### PR TITLE
factor out conf and arg parsing

### PR DESCRIPTION
**Pull request by thewilsonator**

### CODE REVIEW

**Comment by UplinkCoder** [comment]
prepareArguments is a strange name.
Shouldn't it be `parseCommandlineAndConfig` ?

**Comment by thewilsonator** [comment]
done.

File: src/dmd/mars.d [hunk 1]
```diff
@@ -794,6 +694,107 @@ private int tryMain(size_t argc, const(char)** argv, ref Param params)
...
 }
 
+/**
+ * Parses the command line arguments and configuration files
+ */
+version(NoMain) {} else
+int parseCommandlineAndConfig(size_t argc, const(char)** argv, ref Param params, ref Strings files)
+{
+    // Check for malformed input
+    void badArgs()
```
Comment (UplinkCoder) [inline src/dmd/mars.d:704]: this can be static?

File: src/dmd/mars.d [hunk 2]
```diff
@@ -794,6 +694,107 @@ private int tryMain(size_t argc, const(char)** argv, ref Param params)
     return status;
 }
 
+/**
+ * Parses the command line arguments and configuration files
+ */
+version(NoMain) {} else
+int parseCommandlineAndConfig(size_t argc, const(char)** argv, ref Param params, ref Strings files)
```
Comment (UplinkCoder) [inline src/dmd/mars.d:701]: why does it return int instead of bool?
can the ref parameters be out parameters instead?

Comment (thewilsonator) [reply src/dmd/mars.d]: fixed

**Comment by UplinkCoder** [comment]
Looks like you need to do that before parsing
    if (params.color)
        global.console = Console.create(core.stdc.stdio.stderr);

**Comment by thewilsonator** [comment]
But it does do that, thats what I don't understand.

**Comment by thewilsonator** [comment]
Ahh those params can't be `out`, thats what the problem is.

**Comment by thewilsonator** [comment]
@UplinkCoder this is good to go.

**Review by UplinkCoder — APPROVED** [review]
Looks good to me

**Comment by UplinkCoder** [comment]
I'll leave others some time to comment

File: src/dmd/mars.d [hunk 3]
```diff
@@ -794,6 +694,114 @@ private int tryMain(size_t argc, const(char)** argv, ref Param params)
     return status;
 }
 
+/**
+ * Parses the command line arguments and configuration files
```
Comment (WalterBright) [inline src/dmd/mars.d:698]: Missing `Params:` and `Returns:` section

File: src/dmd/mars.d [hunk 4]
```diff
@@ -794,6 +694,114 @@ private int tryMain(size_t argc, const(char)** argv, ref Param params)
...
+ */
+version(NoMain) {} else
+bool parseCommandlineAndConfig(size_t argc, const(char)** argv, ref Param params, ref Strings files)
+{
+    // Check for malformed input
```
Comment (WalterBright) [inline src/dmd/mars.d:703]: This function isn't checking, it's handling the failure.

File: src/dmd/mars.d [hunk 5]
```diff
@@ -794,6 +694,121 @@ private int tryMain(size_t argc, const(char)** argv, ref Param params)
...
 
+/**
+ * Parses the command line arguments and configuration files
+ *
+ * Params:
+ *   argc = Number of arguments passed via command line
+ *   argv = Array of string arguments passed via command line
+ *   params = parametes from argv
+ *   files = files from argv
+ *  Returns: true on faiure
```
Comment (RazvanN7) [inline src/dmd/mars.d:705]:
```suggestion
 * Returns: true on faiure
```

Comment (RazvanN7) [reply src/dmd/mars.d:705]: Usually, the style is to have a newline between the last parameter and the `Returns`.

File: src/dmd/mars.d [hunk 6]
```diff
@@ -794,6 +694,121 @@ private int tryMain(size_t argc, const(char)** argv, ref Param params)
...
+    global.inifilename = parse_conf_arg(&arguments);
+    if (global.inifilename)
+    {
+        // can be empty as in -conf=
+        if (global.inifilename.length && !FileName.exists(global.inifilename))
+            error(Loc.initial, "Config file '%.*s' does not exist.",
+                  cast(int)global.inifilename.length, global.inifilename.ptr);
+    }
+    else
+        global.inifilename = findConfFile(params.argv0, iniName);
```
Comment (RazvanN7) [inline src/dmd/mars.d:757]: DMD style is to use brackets for the `else` branch also, if the `if` branch does, even if it's only one line.

**Pull request merged**